cpp/src_prims/stats/histogram.h

```diff
@@ -0,0 +1,468 @@
+      op(binId, i + j, col);
+    }
+  }
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+__global__ void gmemHistKernel(int* bins, const DataT* data, IdxT nrows,
+                               IdxT nbins, BinnerOp binner) {
+  auto op = [=] __device__(int binId, IdxT row, IdxT col) {
+    if (row >= nrows) return;
+    auto binOffset = col * nbins;
+#if __CUDA_ARCH__ < 700
+    atomicAdd(bins + binOffset + binId, 1);
+#else
+    auto amask = __activemask();
+    auto mask = __match_any_sync(amask, binId);
+    auto leader = __ffs(mask) - 1;
+    if (laneId() == leader) {
+      atomicAdd(bins + binOffset + binId, __popc(mask));
+    }
+#endif  // __CUDA_ARCH__
+  };
+  histCoreOp<DataT, BinnerOp, IdxT, VecLen>(data, nrows, nbins, binner, op,
+                                            blockIdx.y);
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+void gmemHist(int* bins, IdxT nbins, const DataT* data, IdxT nrows, IdxT ncols,
+              BinnerOp binner, cudaStream_t stream) {
+  auto blks = computeGridDim<IdxT, VecLen>(nrows, ncols);
+  gmemHistKernel<DataT, BinnerOp, IdxT, VecLen>
+    <<<blks, ThreadsPerBlock, 0, stream>>>(bins, data, nrows, nbins, binner);
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+__global__ void smemHistKernel(int* bins, const DataT* data, IdxT nrows,
+                               IdxT nbins, BinnerOp binner) {
+  extern __shared__ unsigned sbins[];
+  for (auto i = threadIdx.x; i < nbins; i += blockDim.x) {
+    sbins[i] = 0;
+  }
+  __syncthreads();
+  auto op = [=] __device__(int binId, IdxT row, IdxT col) {
+    if (row >= nrows) return;
+#if __CUDA_ARCH__ < 700
+    atomicAdd(sbins + binId, 1);
+#else
+    auto amask = __activemask();
+    auto mask = __match_any_sync(amask, binId);
+    auto leader = __ffs(mask) - 1;
+    if (laneId() == leader) {
+      atomicAdd(sbins + binId, __popc(mask));
+    }
+#endif  // __CUDA_ARCH__
+  };
+  IdxT col = blockIdx.y;
+  histCoreOp<DataT, BinnerOp, IdxT, VecLen>(data, nrows, nbins, binner, op,
+                                            col);
+  __syncthreads();
+  auto binOffset = col * nbins;
+  for (auto i = threadIdx.x; i < nbins; i += blockDim.x) {
+    atomicAdd(bins + binOffset + i, sbins[i]);
+  }
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+void smemHist(int* bins, IdxT nbins, const DataT* data, IdxT nrows, IdxT ncols,
+              BinnerOp binner, cudaStream_t stream) {
+  auto blks = computeGridDim<IdxT, VecLen>(nrows, ncols);
+  size_t smemSize = nbins * sizeof(unsigned);
+  smemHistKernel<DataT, BinnerOp, IdxT, VecLen>
+    <<<blks, ThreadsPerBlock, smemSize, stream>>>(bins, data, nrows, nbins,
+                                                  binner);
+}
+
+template <unsigned BIN_BITS>
+DI void incrementBin(unsigned* sbins, int* bins, int nbins, int binId,
+                     int col) {
```

Comment:
Is it possible to drop the `col` parameter, and pass `bins + col * nbins` as `bins` instead?

The same applies to `incrementBin<1>`.

Reply:
The catch is for computing `iword` you'd need `binId`, and when the smem-bin overflows, we'd need `col` and `binId` to know where to update the overflowed bin. If we just pass `bins`, we'd have to use modulo operation to know the `binId`. Thus, explicitly passing `col` as an argument.

Reply:
My point is to replace calls to `incrementBin<Bits::BIN_BITS>(sbins, bins, (int)nbins, binId, col)` with `incrementBin<Bits::BIN_BITS>(sbins, bins + col * nbins, (int)nbins, binId)`, and to replace `bins + binOffset` within `incrementBin()` with just `bins`. `binId` would remain the same.

Reply:
Done.